src/rgw/rgw_s3_mirror_sync_tasks.h

```diff
@@ -0,0 +1,181 @@
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2020-2021 IBM Research Europe <koutsovasilis.panagiotis1@ibm.com>
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License version 2, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
+
+#include <mutex>
+#include <shared_mutex>
+#include <string>
+#include <vector>
+#include <map>
+
+#include "rgw_sal.h"
+#include "common/ceph_time.h"
+
+#include <aws/core/Aws.h>
+
+// Base S3MirrorTask that is used only for synchronization between 
+// other tasks
+class S3MirrorTask
+{
+private:
+  std::mutex task_mutex;
+  std::condition_variable condVar;
+
+public:
+    bool completed = false;
+    int ret = 0;
+    
+    S3MirrorTask() {};
+    ~S3MirrorTask() {};
+
+    int wait() {
+        std::unique_lock<std::mutex> lck(task_mutex);
+        condVar.wait(lck, [this]{ return this->completed; });
+        return ret;
+    }
+
+    int waitfor(std::chrono::milliseconds dur, bool& cv_status) {
+        std::unique_lock<std::mutex> lck(task_mutex);
+        cv_status = condVar.wait_for(lck, dur, [this]{ return this->completed; });
+        return ret;
+    }
+
+    void notify(int ret) {
+        {
+            std::lock_guard<std::mutex> lck(task_mutex);
+            completed = true;
+            this->ret = ret;
+        }
+        condVar.notify_all();
+    }
+};
+
+// S3MirrorDownloadTask inherits S3MirrorTask and carries further variables 
+// that regard downloading an object through AWS S3 client
+class S3MirrorDownloadTask : public S3MirrorTask, public std::enable_shared_from_this<S3MirrorTask>
+{
+public:
+  class CustomStreamBuffer : public std::streambuf
+  {
+    private:
+      std::unique_ptr<char[]> bodybuffer;
+      long long length =0;
+    public:
+      CustomStreamBuffer(long long size) 
+      {
+        
+        bodybuffer.reset(new char[size]);
+
+        this->pubsetbuf(&(bodybuffer[0]), size);
+        this->setg(0, 0, 0);
+        this->setp(&(bodybuffer[0]), &(bodybuffer[size-1]));
+      }
+      
+      streamsize xsputn (const char* s, streamsize n) override {
+        memcpy(&(bodybuffer[length]), s, n);
+        length += n;
+        return n;
+      }
+
+      char *get_buffer() {
+        return &(bodybuffer[0]);
+      }
+
+      long long get_unsafe_length()
+      {
+        return length;
+      }
+  };
+  // necessary to write the object locally
+  rgw_user user_id;
+  // object hotness ref counter
+  std::atomic<int> refs = 0;
+  // vector with result to keep the iostream alive
+  std::vector<Aws::S3::Model::GetObjectResult> result;
+  std::shared_ptr<CustomStreamBuffer> customStreamBuf = nullptr;
+  std::basic_iostream<char, std::char_traits<char>>  *ioStream;
+  // S3 object properties to dump the headers
+  ceph::time_detail::real_clock::time_point expires;
```

Comment:
```suggestion
real_clock::time_point expires;
```
fix compilation error (fedora 32/clang 10):
```
FAILED: src/rgw/CMakeFiles/rgw_common.dir/rgw_op.cc.o
...
In file included from ../src/rgw/rgw_op.cc:77:
In file included from ../src/rgw/rgw_s3_mirror.h:46:
../src/rgw/rgw_s3_mirror_sync_tasks.h:108:3: error: no member named 'real_clock' in namespace 'ceph::time_detail'; did you mean simply 'real_clock'?
  ceph::time_detail::real_clock::time_point expires;
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  real_clock
../src/common/ceph_time.h:72:7: note: 'real_clock' declared here
class real_clock {
      ^
```

Reply:
I am not sure why my gcc on debian never complained about that, but thanks for letting me know